src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala

```diff
@@ -65,6 +65,31 @@ object ArrayFunctions extends RegistryFunctions {
     ArrayFold(a, one, product, v, If(IsNA(Ref(v, t)), Ref(product, t), ApplyBinaryPrimOp(Multiply(), Ref(product, t), Ref(v, t))))
   }
 
+  def median(array: IR): IR = {
+    val t = -array.typ.asInstanceOf[TArray].elementType
+    val a = Ref(genUID(), TArray(t))
+    val size = Ref(genUID(), TInt32())
+    val lastIdx = size - 1
+    val midIdx = lastIdx.floorDiv(2)
+    def ref(i: IR) = ArrayRef(a, i)
+    val len: IR = ArrayLen(a)
+    def div(a: IR, b: IR): IR = ApplyBinaryPrimOp(BinaryOp.defaultDivideOp(t), a, b)
+
+    Let(a.name, ToArray(array),
+      If(IsNA(a),
+        NA(t),
+        Let(size.name,
+          If(len.ceq(0), len, If(IsNA(ref(len - 1)), len - 1, len)),
```

Comment:
I think this doesn't generalize quite like this because for the Array case, you'll probably want to just filter out all the `NA`s first, in which case this check is redundant. (you also probably don't need the `ToArray` above since you're already calling it on the set.